src/main/java/com/hazelcast/gcp/GcpConnectionException.java

```diff
@@ -0,0 +1,125 @@
+import com.hazelcast.logging.Logger;
+
+/**
+ * Exception to indicate an issue when pinging GCP API.
+ *
+ * A list of errors can be found at: {@see https://cloud.google.com/resource-manager/docs/core_errors}.
+ */
+
+public class GcpConnectionException extends RuntimeException {
+    /**
+     * Static error string returned when there are insufficient permissions to access the Compute Engine via the API
+     */
+    public static final String S_GCP_ERROR_INSUFFICIENT_PERMISSION_SCOPE = "Insufficient Permission: "
+            + "Request had insufficient authentication scopes.";
+    /**
+     * Static error string returned when service account doesn't have IAM role to access 'compute.instances.list'
+     */
+    public static final String S_GCP_ERROR_COMPUTE_INSTANCES_LIST = "Required 'compute.instances.list' permission for";
+
+    private static final ILogger LOGGER = Logger.getLogger(GcpConnectionException.class);
+
+    private String domain;
+    private String reason;
+    private String gcpMessage;
+    private int errorCode;
+    private String url;
+
+    public GcpConnectionException(String jsonResponse) {
+        super("Failure in executing GCP API request");
+
+        setFromJson(jsonResponse);
+    }
+
+    public GcpConnectionException(String jsonResponse, String url) {
+        super("Failure in executing GCP API request at: " + url);
+        this.url = url;
+
+        setFromJson(jsonResponse);
+    }
+
+    private void setFromJson(String jsonResponse) {
+        JsonObject gcpErrorJson = Json.parse(jsonResponse).asObject().get("error").asObject();
+        JsonArray gcpErrorList = toJsonArray(gcpErrorJson.get("errors"));
+        JsonObject gcpError = gcpErrorList.get(0).asObject();
+        this.domain = gcpError.get("domain").asString();
+        this.reason = gcpError.get("reason").asString();
+        this.gcpMessage = gcpError.get("message").asString();
+        this.errorCode = gcpErrorJson.get("code").asInt();
+    }
+
+    public String getDomain() {
+        return this.domain;
+    }
+    public String getReason() {
+        return this.reason;
+    }
+    public String getGcpMessage() {
+        return this.gcpMessage;
+    }
+    public int getErrorCode() {
+        return this.errorCode;
+    }
+    public String getUrl() {
+        return this.url;
+    }
+
+    public static Boolean getIsGcpConnectionException(String responseString) {
+        JsonValue gcpErrorValue = Json.parse(responseString).asObject().get("error");
+        if (gcpErrorValue != null) {
+            JsonArray gcpErrorList = toJsonArray(gcpErrorValue.asObject().get("errors"));
+            if (!gcpErrorList.isEmpty()) {
+                String gcpErrorMessage = gcpErrorList.get(0).asObject().get("message").asString();
+                if (gcpErrorMessage != null) {
+                    return Boolean.TRUE;
+                }
+            }
+        }
+        return Boolean.FALSE;
+    }
+
+    public void logErrorAndThrowException() {
```

Comment:
I don't think it's the right place for logging. Could you move it outside the exception class?

You can prepare the message in the exception class, so it returns `getMessage()` with a nice message. But do not use LOGGER here.